File: kafka/src/test/groovy/io/micronaut/configuration/kafka/docs/consumer/sendto/ProductListener.java
package io.micronaut.configuration.kafka.docs.consumer.sendto;

// tag::imports[]
import io.micronaut.configuration.kafka.annotation.KafkaKey;
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.configuration.kafka.docs.consumer.config.Product;
import io.micronaut.messaging.annotation.SendTo;
import io.reactivex.Single;
// end::imports[]

@KafkaListener
public class ProductListener {

    // tag::method[]
    @Topic("awesome-products") // <1>
    @SendTo("product-quantities") // <2>
    public int receive(@KafkaKey String brand,
                       Product product) {
        System.out.println("Got Product - " + product.getName() + " by " + brand);

        return product.getQuantity(); // <3>
    }
    // end::method[]

    // tag::reactive[]
    @Topic("awesome-products") // <1>
    @SendTo("product-quantities") // <2>
    public Single<Integer> receiveProduct(@KafkaKey String brand,
                                          Single<Product> productSingle) {

        return productSingle.map(product -> {
            System.out.println("Got Product - " + product.getName() + " by " + brand);
            return product.getQuantity(); // <3>
        });
    }
    // end::reactive[]
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/reactive/ProductListener.py
import logging
from typing import Annotated

from micronaut.configuration.kafka.annotation import KafkaKey, KafkaListener, Topic
from micronaut.context.annotation import Requires
from micronaut.core.annotation import Blocking
from micronaut.kafka.docs.Product import Product
from reactor.core.publisher import Mono

LOG = logging.getLogger(__name__)


@Requires(property="spec.name", value="ReactiveProductListenerTest")
@KafkaListener
class ProductListener:

    # tag::method[]
    @Topic("reactive-products")
    def receive(self, brand: Annotated[str, KafkaKey],  # <1>
                product_publisher: Mono[Product]) -> Mono[Product]:  # <2>
        return product_publisher.doOnSuccess(
            lambda product: LOG.info("Got Product - %s by %s", product.name, brand)  # <3>
        )
    # end::method[]

    # tag::blocking[]
    @Blocking
    @Topic("reactive-products")
    def receive_blocking(self, brand: Annotated[str, KafkaKey], product_publisher: Mono[Product]) -> Mono[Product]:
        return product_publisher.doOnSuccess(
            lambda product: LOG.info("Got Product - %s by %s", product.name, brand)
        )
    # end::blocking[]
